lib/luacrypto.c

```diff
@@ -0,0 +1,438 @@
+	if (iv_len != iv_len_expected)
+		return luaL_error(
+			L, "incorrect nonce length: expected %zu, got %zu",
+			iv_len_expected, iv_len
+		);
+
+	authsize = crypto_aead_authsize(tfm);
+	if (ct_len < authsize)
+		return luaL_error(L, "ciphertext too short (must include tag)");
+
+	pt_len = ct_len - authsize;
+
+	iv_buf = kmemdup(iv, iv_len, gfp);
+	ct_buf = kmemdup(ciphertext, ct_len, gfp);
+	if (aad_len > 0) {
+		aad_buf = kmemdup(aad, aad_len, gfp);
+	} else {
+		aad_buf = NULL;
+	}
+	if (!ct_buf || !iv_buf || (aad_len > 0 && !aad_buf)) {
+		kfree(aad_buf);
+		kfree(ct_buf);
+		kfree(iv_buf);
+		return luaL_error(L, "failed to allocate buffers");
+	}
+
+	req = aead_request_alloc(tfm, gfp);
+	if (!req) {
+		kfree(aad_buf);
+		kfree(ct_buf);
+		kfree(iv_buf);
+		return luaL_error(L, "failed to allocate aead request");
+	}
+
+	if (aad_len > 0) {
+		sg_init_table(sg, 2);
+		sg_set_buf(&sg[0], aad_buf, aad_len);
+		sg_set_buf(&sg[1], ct_buf, ct_len);
+	} else {
+		sg_init_table(sg, 1);
+		sg_set_buf(&sg[0], ct_buf, ct_len);
+	}
+
+	aead_request_set_ad(req, aad_len);
+	aead_request_set_callback(req, 0, NULL, NULL);
+	aead_request_set_crypt(req, sg, sg, pt_len, iv_buf);
+
+	ret = crypto_aead_decrypt(req);
+
+	aead_request_free(req);
+	kfree(aad_buf);
+	kfree(iv_buf);
+
+	if (ret) { /* Decryption or authentication failed */
+		kfree(ct_buf);
+		return luaL_error(L, "decryption failed (err %d, possibly auth error)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)ct_buf, pt_len);
+	kfree(ct_buf);
+	return 1;
+}
+
+/*
+* Release callback: Called when the Lua crypto object is garbage collected.
+* Frees the crypto transform.
+*/
+static void luacrypto_release(void *private)
+{
+	struct crypto_aead *tfm = (struct crypto_aead *)private;
+	if (tfm)
+	crypto_free_aead(tfm);
+}
+
+/*
+* Lua object methods for the crypto API.
+*/
+static const luaL_Reg luacrypto_mt[] = {
+	{"setkey", luacrypto_setkey},
+	{"setauthsize", luacrypto_setauthsize},
+	{"encrypt", luacrypto_encrypt},
+	{"decrypt", luacrypto_decrypt},
+	{"close", lunatik_closeobject},
+	{"__gc", lunatik_deleteobject},
+	{"__close", lunatik_closeobject},
+	{"__index", lunatik_monitorobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luacrypto_class = {
+	.name = "crypto",
+	.methods = luacrypto_mt,
+	.release = luacrypto_release,
+	.sleep = true,
+	.pointer = true,
+};
+
+/*
+* crypto.new(algname)
+*
+* Creates a new crypto object. The argument algname is a string, e.g., "gcm(aes)",
+* that specifies the algorithm. This design is generic so that additional algorithms
+* and modes can be supported.
+*/
+static int luacrypto_new(lua_State *L)
+{
+	const char *algname = luaL_checkstring(L, 1);
+	lunatik_object_t *object;
+	struct crypto_aead *tfm;
+	tfm = crypto_alloc_aead(algname, 0, 0);
+	if (IS_ERR(tfm))
+		luaL_error(
+			L, "failed to allocate transform for %s (%ld)",
+			algname, PTR_ERR(tfm)
+		);
+
+	object = lunatik_newobject(L, &luacrypto_class, 0);
```

Comment:
I would do this just after checking string, otherwise, if it fails to allocate mem, tfm will leak, because the object will never be create and therefore, it will never be released..